### PR TITLE
Use implicit login, retry failed HSM operation.

### PR DESCRIPTION
**Pull request by DanielChicot**

### CODE REVIEW

File: src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java [hunk 1]
```diff
@@ -49,41 +50,41 @@ public void generateDataKey() {
...
-    public void handlesKMSInternalException() {
+    public void handlesKMSInternalException() throws MasterKeystoreException {
         throwException(KMSInternalException.class);
     }
 
     @Test(expected = DataKeyGenerationException.class)
-    public void handlesKMSInvalidStateException() {
+    public void handlesKMSInvalidStateException() throws MasterKeystoreException {
         throwException(KMSInvalidStateException.class);
     }
```
Comment (mark-r-m) [inline src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java]: These test names don't tell us _why_ we are testing and _what_ they are testing. It looks just like Inheritance tree tests?

File: src/test/java/uk/gov/dwp/dataworks/services/DataKeyServiceTest.java [hunk 2]
```diff
@@ -32,7 +34,7 @@ public void init() {
         Mockito.reset(currentKeyIdProvider);
         //hsmCredentialsProvider.clearCache();
```
Comment (mark-r-m) [inline src/test/java/uk/gov/dwp/dataworks/services/DataKeyServiceTest.java]: Can the comment be removed?

File: src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java [hunk 3]
```diff
@@ -49,41 +50,41 @@ public void generateDataKey() {
...
 
-    private void throwException(Class<? extends Exception> e) throws CurrentKeyIdException {
+    private void throwException(Class<? extends Exception> e) throws CurrentKeyIdException, MasterKeystoreException {
```
Comment (mark-r-m) [inline src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java]: Is this testing the various catch blocks and their content?

File: src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java (hunk 1 above)
Comment (DanielChicot) [reply src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java]: These are the KMS tests - they have not been introduced as part of this PR - the only changes are updates to the signature to accommodate the new exception introduced. If we want to retrofit all these tests could it be done in a tech debt ticket?

File: src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java (hunk 3 above)
Comment (DanielChicot) [reply src/test/java/uk/gov/dwp/dataworks/provider/kms/KMSDataKeyGeneratorProviderTest.java]: That's just a utility method that throws an exception. It's not new though and has not been introduced as part of this work - the signature had to be altered to accommodate a new exception.

**Pull request merged**